UTF-16 be, UTF-32 be support

borzacchiello opened the pull request:
<!-- Filling this template is mandatory -->

Hi! I'm trying to become more comfortable with the code trying to solve some issues. I'm sorry if I'm doing something wrong!

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

This PR:
-  refactor `string_scan_range` of `librz/bin/bfile.c` by moving the string searching logic  to the file `librz/util/str_search.c` (as suggested in #1052).
- adds support for the detection of UTF-16 BE and UTF-32 BE strings.

---

The introduced public API due to refactoring is the following:
```C
RZ_API int rz_scan_strings(RzList *list, RzBuffer *buf_to_scan,
	const ut64 from, const ut64 to, int min_str_length, int type);
```
it looks for strings in the `RzBuffer buf_to_scan`, filling the `RzList list`.
`string_scan_range` now uses this API. Unfortunately, it is possibly slower than the previous version, since it always allocates a list that is passed to `rz_scan_strings`. Ofc I can revert these modifications.

---

`UTF-16 BE` and `UTF-32 BE` detection heuristic should be very similar to the one for `UTF-16 LE` and `UTF-32 LE`. It simply checks the following pattern for the first 6 bytes of the string:
UTF-32 BE: 00 00 00 _XX_ 00 00
UTF-16 BE: 00 _XX_ 00 _XX_ 00 _XX_ 

where _XX_ is a byte != 0

---
I did not add (yet) big-endian strings [here](https://github.com/rizinorg/rizin/blob/413c8e2ad789e863fda2a9825680868c4f37ec74/librz/util/str_search.c#L187) since I am not sure that `rz_utf_block_list` would work with big-endian UTF strings.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

I did not add any tests yet. I'm just using [this file](https://github.com/rizinorg/rizin/files/7060773/encodings.zip) to perform some preliminary tests, running:
`rz-bin -zz /path/to/encodings`

output before this commit:
```
[Strings]
nth paddr      vaddr      len size section type    string
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――
0   0x00000000 0x00000000 20  21           ascii   i am an ascii string
1   0x00000020 0x00000020 19  21           utf8    i am a ÙTF-8 string blocks=Basic Latin,Latin-1 Supplement
2   0x00000040 0x00000040 22  46           utf16le i am a UTF-16le string
3   0x00000071 0x00000071 21  43           utf16le i am a UTF-16be strin
4   0x000000a0 0x000000a0 22  92           utf32le i am a UTF-32le string
5   0x00000103 0x00000103 21  88           utf32le i am a UTF-32be strin
```

output after this commit:
```
[Strings]
nth paddr      vaddr      len size section type    string                                                    
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
0   0x00000000 0x00000000 20  21           ascii   i am an ascii string
1   0x00000020 0x00000020 19  21           utf8    i am a ÙTF-8 string blocks=Basic Latin,Latin-1 Supplement
2   0x00000040 0x00000040 22  46           utf16le i am a UTF-16le string
3   0x00000070 0x00000070 22  46           utf16be i am a UTF-16be string
4   0x000000a0 0x000000a0 22  92           utf32le i am a UTF-32le string
5   0x00000100 0x00000100 22  92           utf32be i am a UTF-32be string
```

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

closes #1052

borzacchiello committed:
Refactored string search

string search logic moved from librz/bin/bfile.c to
librz/util/str_search.c

borzacchiello committed:
UTF-16BE and UTF-32BE support

ret2libc reviewed (commented):
First: thanks for your contributions! They are very welcome ;)

Second: I just had a very quick look and I noticed you moved some stuff in librz/util, but you are using rz_bin.h and some RzBin stuff in str_search.c. This should not be done, because ideally librz/util should not depend on RzBin. RzBin depends on RzUtil. Not the other way around.

I will look more at the code later, thanks again.

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+#include <rz_bin.h>
```
You can safely remove this line

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+#define RZ_STRING_SCAN_BUFFER_SIZE 2048
+#define RZ_STRING_MAX_UNI_BLOCKS   4
+
+RZ_API int rz_scan_strings(RzList *list, RzBuffer *buf_to_scan,
```
Please add Doxygen

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+#include <rz_util/rz_assert.h>
+
+// maybe too big sometimes? 2KB of stack eaten here..
+#define RZ_STRING_SCAN_BUFFER_SIZE 2048
```
I suggest to make a structure something like:
```c
typedef struct {
   size_t buf_size;
   size_t max_uni_blocks;
   size_t min_str_length;
} RzUtilStrScanOptions;
```
Moreover, in the future we could provide setting these options from the `RzConfig` options like some of the `e bin.str.*` configuration variables.

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+#define RZ_STRING_MAX_UNI_BLOCKS   4
+
+RZ_API int rz_scan_strings(RzList *list, RzBuffer *buf_to_scan,
+	const ut64 from, const ut64 to, int min_str_length, int type) {
```
I suggest to convert `type` from `int` to the `enum`, strictly limiting the possible values.

borzacchiello replied:
Done

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+				}
+				rc = rz_utf8_encode(tmp + i, r);
+				runes++;
+				/* Print the escape code */
```
We don't "print" anything here.

XVilka replied:
Also probably this part of the function is better as a separate small static function probably.

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+			// back up past the \0 to the last char just in case it starts a wide string
+			needle -= 2;
+		}
+		if (runes >= min_str_length) {
```
This part of the function could also be split in a separate function as well.

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+			bs->size = needle - str_start;
+			bs->ordinal = count++;
+			// TODO: move into adjust_offset
+			switch (str_type) {
```
I would split this part into a separate function as well.

borzacchiello replied:
done

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+	int count = 0, i, rc, runes;
+	int str_type = RZ_STRING_TYPE_DETECT;
+
+	// list cannot be NULL
```
Move these checks before defining variables.

borzacchiello replied:
Done

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+	// buf_to_scan cannot be NULL
+	rz_return_val_if_fail(buf_to_scan, -1);
+
+	if (type == -1) {
```
This can be removed after switching to enum

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+	if (type == -1) {
+		type = RZ_STRING_TYPE_DETECT;
+	}
+	if (from == to) {
```
Move those after `rz_return_val_if_fail()` at the beginning of the function, before defining variables.

borzacchiello replied:
Done

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,271 @@
+		return 0;
+	}
+	if (from > to) {
+		eprintf("Invalid range to find strings 0x%" PFMT64x " .. 0x%" PFMT64x "\n", from, to);
```
Make it `RZ_LOG_ERROR()` probably.

XVilka commented on librz/bin/bfile.c:
```diff
@@ -112,255 +109,70 @@ static void print_string(RzBinFile *bf, RzBinString *string, int raw, PJ *pj) {
+	RzBinString *bs;
+	RzBinSection *s = NULL;
+
+	RzList *str_list = rz_list_new();
```
Please check if it's `NULL`.

borzacchiello replied:
Done

XVilka commented on librz/bin/bin.c:
```diff
@@ -71,11 +71,13 @@ RZ_API RzBinXtrData *rz_bin_xtrdata_new(RzBuffer *buf, ut64 offset, ut64 size, u
 
 RZ_API const char *rz_bin_string_type(int type) {
```
Convert `type` to the enum here as well.

XVilka replied:
Also please add `RZ_OWN` for the return value.

borzacchiello replied:
Done

XVilka commented on librz/bin/bfile.c:
```diff
@@ -112,255 +109,70 @@ static void print_string(RzBinFile *bf, RzBinString *string, int raw, PJ *pj) {
+
+	RzList *str_list = rz_list_new();
+
+	int count = rz_scan_strings(str_list, bf->buf, from, to, min, type);
```
Please convert this `type` to the enum as well.

borzacchiello replied:
Done, I used the already defined enum `RzStrEnc`

XVilka commented:
Good job overall. I suggest to add some unit tests in the new `test/unit/test_str_scan.c`. Now, with the better API you could easily make a good set of the tests.
See these as an example:
- `test/unit/test_str.c`
- `test/unit/test_strbuf.c`
- `test/unit/test_buf.c`

You could use Rizin commands like `pc`, `pcs`, `px` to get the byte data for such unit tests from the binaries.

borzacchiello committed:
Decoupled rz_str_search with rz_bin

XVilka commented on librz/bin/bin.c:
```diff
@@ -69,13 +69,15 @@ RZ_API RzBinXtrData *rz_bin_xtrdata_new(RzBuffer *buf, ut64 offset, ut64 size, u
 }
 
-RZ_API const char *rz_bin_string_type(int type) {
+RZ_API RZ_OWN const char *rz_bin_string_type(int type) {
```
It should be `RZ_BORROW`, sorry.

borzacchiello replied:
done

XVilka commented on librz/include/rz_bin.h:
```diff
@@ -161,6 +161,8 @@ enum {
 	RZ_STRING_TYPE_UTF8 = 'u',
 	RZ_STRING_TYPE_WIDE = 'w', // utf16 / widechar string
```
While we are on it, I wonder if we should make it clear that those `RZ_STRING_TYPE_WIDE*` are in fact `WIDE*_LE`. cc @ret2libc @kazarmy @wargio @pelijah

wargio replied:
better to rename it

XVilka commented on librz/include/rz_bin.h:
```diff
@@ -899,7 +901,7 @@ RZ_API void rz_bin_list_archs(RzBin *bin, PJ *pj, int mode);
 RZ_API RzBuffer *rz_bin_package(RzBin *bin, const char *type, const char *file, RzList *files);
 
-RZ_API const char *rz_bin_string_type(int type);
+RZ_API RZ_OWN const char *rz_bin_string_type(int type);
```
`RZ_BORROW`

borzacchiello replied:
done

XVilka replied:
You forgot to change this one.

borzacchiello replied:
Ups, fixed

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,280 @@
+RZ_API int rz_scan_strings(RzList *list, RzBuffer *buf_to_scan,
+	const ut64 from, const ut64 to, int min_str_length, RzStrEnc type) {
+
+	// list cannot be NULL
```
No need for these comments in both cases.

borzacchiello replied:
Removed

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,280 @@
+	}
+}
+
+RZ_API int rz_scan_strings(RzList *list, RzBuffer *buf_to_scan,
```
You added the structure for options, but you don't use it as a function parameter. Please do not hardcode it (at least in RzUtil) and pass it as a parameter.

borzacchiello replied:
done! Now it is the first parameter

ret2libc commented on librz/include/rz_util/rz_str_search.h:
```diff
@@ -0,0 +1,36 @@
+extern "C" {
+#endif
+
+typedef struct {
```
use doxygen style for these. See rz_cmd.h for an example on how to do it.

borzacchiello replied:
done! I hope that I made it correctly

ret2libc commented on librz/include/rz_util/rz_str_search.h:
```diff
@@ -0,0 +1,36 @@
+} RzUtilStrScanOptions;
+
+RZ_API void rz_detected_string_free(RzDetectedString *str);
+
```
```suggestion
```

ret2libc commented on librz/include/rz_util/rz_str_search.h:
```diff
@@ -0,0 +1,36 @@
+	RzStrEnc type;
+} RzDetectedString;
+
+typedef struct {
+	size_t buf_size;
+	size_t max_uni_blocks;
+	size_t min_str_length;
+} RzUtilStrScanOptions;
```
I think there's no need for this to be here. It seems you are just using this inside str_search.c, so you could probably just move this structure in that file.

borzacchiello replied:
now RzUtilStrScanOptions is a parameter of `rz_scan_strings`

ret2libc commented on librz/bin/bfile.c:
```diff
@@ -110,257 +107,118 @@ static void print_string(RzBinFile *bf, RzBinString *string, int raw, PJ *pj) {
+		break;
+	}
+
+	dst->string = rz_str_ndup(src->string, src->size);
```
I think inside this function we should free the `src` string and re-use the src->string from there instead of duplicating it. There are already a lot of memory allocations in Rizin and many of them can probably be avoided.

borzacchiello replied:
Modified

XVilka commented:
@borzacchiello, do you plan to continue the development of this PR?

borzacchiello commented:
@XVilka Yes, sorry for the delay! I was a little busy last week. Tomorrow I will work on it!

XVilka commented:
Please rebase to resolve the conflicts

XVilka reviewed (changes requested):
Looks very good now. But in the main API function, please put the `RzBuffer` parameter that we are going to scan first, before everything else.

borzacchiello reviewed (commented):
I should have addressed all your comments. I also added some tests in `test/unit/test_str_search.c`.

Unfortunately, I think that I broke something since I see some related errors in Travis. I will look into them!

I still need to extend `reduce_false_positives` to big-endian strings.

borzacchiello committed:
Changed parameter order in rz_scan_strings

borzacchiello commented:
> Looks very good now. But in the main API function, please put the `RzBuffer` parameter that we are going to scan first, before everything else.

Done

borzacchiello commented:
> Please rebase to resolve the conflicts

Done (I think)

XVilka commented:
There are failed tests like those:
```


[XX] db/cmd/cmd_i ascii substring detection (#14499)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc izz~pomf bins/pe/Reborn_Stub-strings.exe
-- stdout
--- expected
+++ actual
@@ -1,4 +1,4 @@
-7154 0x00087f8a 0x00489d8a 26  53   .text   utf16le http://pomf.cat/upload.php
-7159 0x000880dd 0x00489edd 19  39   .text   utf16le https://a.pomf.cat/
-7322 0x00089bed 0x0048b9ed 26  53   .text   utf16le http://pomf.cat/upload.php
-7323 0x00089c22 0x0048ba22 19  40   .text   utf16le https://a.pomf.cat/
+7192 0x00087f8a 0x00489d8a 26  53   .text   utf16le http://pomf.cat/upload.php
+7197 0x000880dd 0x00489edd 19  39   .text   utf16le https://a.pomf.cat/
+7360 0x00089bed 0x0048b9ed 26  53   .text   utf16le http://pomf.cat/upload.php
+7361 0x00089c22 0x0048ba22 19  40   .text   utf16le https://a.pomf.cat/



[XX] db/cmd/cmd_i izz and bin.str.enc
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'wz k\0i\0abcdefgh
(izzenc enc; e bin.str.enc=$0; izz:quiet)
.(izzenc guess)
.(izzenc ascii)
.(izzenc utf8)
.(izzenc utf16le)
?e
wx 00250025002500250000
.(izzenc guess)
.(izzenc utf16le)
?e
wx 26252625262526250000
.(izzenc guess)
.(izzenc utf16le)
' =
-- stdout
--- expected
+++ actual
@@ -3,7 +3,7 @@
 0   0x00000004 0x00000004 8   9            ascii abcdefgh
 0   0x00000000 0x00000000 6   14           utf16le ki扡摣晥桧 blocks=Basic Latin,CJK Unified Ideographs
 
-0   0x00000001 0x00000001 4   9            utf16le %%%%
+0   0x00000000 0x00000000 4   10           utf16be %%%%
 0   0x00000000 0x00000000 4   10           utf16le ──── blocks=Box Drawing
 
 0   0x00000000 0x00000000 8   9            ascii &%&%&%&%


[XX] db/formats/mdmp mdmp 64bit - strings count
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iz~? bins/mdmp/hello64.dmp
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-22900
+22913

-- stderr
[WARN] Invalid or unsupported enumeration encountered 21
[WARN] Invalid or unsupported enumeration encountered 22
[INFO] Parsing data sections for large dumps can take time, please be patient (but if strings ain't your thing try with -z)!



[XX] db/formats/mdmp mdmp 32bit - strings count
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iz~? bins/mdmp/hello.dmp
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-10276
+10292

-- stderr
[WARN] Invalid or unsupported enumeration encountered 21
[WARN] Invalid or unsupported enumeration encountered 22
Error parsing resource directory
[INFO] Parsing data sections for large dumps can take time, please be patient (but if strings ain't your thing try with -z)!
```
See https://github.com/rizinorg/rizin/pull/1535/checks?check_run_id=3527121864#step:16:111

XVilka commented on librz/util/str_search.c:
```diff
@@ -0,0 +1,301 @@
+	return 0;
+}
+
+RZ_API int rz_scan_strings(RzBuffer *buf_to_scan, RzList *list, const RzUtilStrScanOptions *opt,
```
Please add Doxygen description of the function, it is especially important in regard to the parameters and `opt` description. Also the return values.

borzacchiello replied:
I added the description to the header file, I have to add it here?

XVilka replied:
Yes, please move it from the header to here.

XVilka reviewed (changes requested):
Overall good job - the only two things left - Doxygen and fixing tests.

borzacchiello committed:
Fixed needle reset in rz_string_search

Now when the string search fails, needle is restored to original_needle+1,
i.e., we continue the searching process starting from the next byte.

borzacchiello committed:
Changed heuristic for UTF BE detection

one more byte in for the detection (6 instead of 5)

borzacchiello committed:
Fixed unit test in str_search

XVilka commented:
There are regressions like that:
```
[XX] db/analysis/x86_64 Mixed reg/stack callee arguments type matching
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
s 0x14000b54c
aei
aeim
af
aft
pdf
' bins/pe/testapp-msvc64.exe
-- stdout
--- expected
+++ actual
@@ -5,7 +5,7 @@
 |           ; var HANDLE hTemplateFile @ rsp+0x30
 |           0x14000b54c      4883ec48       sub   rsp, 0x48
 |           0x14000b550      488364243000   and   qword [hTemplateFile], 0 ; HANDLE hTemplateFile
-|           0x14000b556      488d0da37300.  lea   rcx, str.CONOUT      ; 0x140012900 ; u"CONOUT$" ; LPCWSTR lpFileName
+|           0x14000b556      488d0da37300.  lea   rcx, [0x140012900]   ; u"CONOUT$" ; LPCWSTR lpFileName
 |           0x14000b55d      8364242800     and   dword [dwFlagsAndAttributes], 0 ; DWORD dwFlagsAndAttributes
 |           0x14000b562      41b803000000   mov   r8d, 3               ; DWORD dwShareMode
 |           0x14000b568      4533c9         xor   r9d, r9d             ; LPSECURITY_ATTRIBUTES lpSecurityAttributes
```

XVilka commented:
@imbillow for [EBCDIC strings](https://github.com/rizinorg/rizin/pull/1643) search I recommend basing your work on top of this one - this interface is now better suitable for the extension. @borzacchiello any updates on addressing the regressions?

borzacchiello commented:
> @imbillow for [EBCDIC strings](https://github.com/rizinorg/rizin/pull/1643) search I recommend basing your work on top of this one - this interface is now better suitable for the extension. @borzacchiello any updates on addressing the regressions?

I'm working on them! Some depend on the ambiguity between UTF-LE and UTF-BE. For example, [this string](https://github.com/rizinorg/rizin/blob/dev/test/db/cmd/cmd_i#L4473) can be either "%%%%" UTF-16LE, if the string starts at offset 1, or "%%%%" UTF-16BE if the string starts at offset 0.

b1llow commented:
Do I need to wait for it to be done here? @XVilka

XVilka commented:
@imbillow you could make a branch based on this one, then if @borzacchiello makes any updates you could always rebase.

XVilka commented:
@borzacchiello regarding ambiguity, we could edit the tests removing them, for example. Or ignoring these particular cases.

borzacchiello committed:
Refactored rz_scan_strings and better UTF16-LE and UTF16-BE detection

Now if there is an ambiguity between UTF16-LE and UTF16-BE, we take the
longest one.

XVilka reviewed (changes requested):
```
# MISSING COPYRIGHT AND LICENSING INFORMATION

The following files have no copyright and licensing information:
* librz/util/str_search.c
* test/unit/test_str_search.c
```
Please also rebase on top of the latest `dev`

borzacchiello commented:
@XVilka I refactored the code a little bit, now it should be more readable. Furthermore, I changed UTF16LE and BE detection. Now if there is any ambiguity, we will choose the longest one.
There are some tests in which we are counting the number of strings, they will likely fail due to [this commit](https://github.com/rizinorg/rizin/pull/1535/commits/7274d233c53212950397b44c5e32cfee144ce4f9). In the original code, if a string detection failed, `needle` was not reset to `original_needle + 1`, possibly skipping strings

XVilka commented:
@borzacchiello please fix these tests then, carefully comparing the results first. I recommend to use `diff` util to compare the output, posting here the difference would be helpful as well.

borzacchiello committed:
add license

XVilka commented:
What worries me, are these regressions:
```

[XX] db/cmd/cmd_pd bin.str.enc utf32le
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=false
e asm.cmt.off=false
e str.escbslash=true
e bin.str.enc=utf16le
pd 1 @ 0x004016de
e bin.str.enc=utf32le
pd 1 @ 0x004016de
e asm.sub.names=false
e asm.noisy=false
e bin.str.enc=guess
pd 1 @ 0x004016ed
pd 1 @ 0x004016f7
pd 1 @ 0x00401701
e bin.str.enc=utf32le
pd 1 @ 0x00401701
' bins/elf/strenc
-- stdout
--- expected
+++ actual
@@ -1,6 +1,6 @@
-            0x004016de      mov   esi, str.Linux_wide__esc:__e_0m      ; u"\t"
-            0x004016de      mov   esi, str.Linux_wide__esc:__e_0m      ; U"\tLinux_wide\\esc: \x1b[0m\xa1\r\n"
-            0x004016ed      mov   edi, 0x40258c                        ; U"utf32le> \\u00a2\\u20ac\\U00010348 in cyan:\x1b[36m \xa2\u20ac\U00010348 \x1b[0m\n"
+            0x004016de      mov   esi, 0x402528                        ; '(%@' ; u"\t"
+            0x004016de      mov   esi, 0x402528                        ; '(%@' ; U"\tLinux_wide\\esc: \x1b[0m\xa1\r\n"
+            0x004016ed      mov   edi, 0x40258c
@@ -3,5 +3,5 @@
             0x00401693      mov   edi, 0x4021ff                        ; "utf8> \\u00a2\\u20ac\\U00010348 in yellow:.[33m ......... .[0m\n"
             0x00401693      mov   edi, 0x4021ff                        ; "utf8> \\u00a2\\u20ac\\U00010348 in yellow:.[33m ... .[0m."
             0x004016ac      mov   edi, 0x40224a                        ; 'J"@' ; u"utf16le> \\u00a2\\u20ac\\U00010348 in green:.[32m ... .[0m."
-            0x004016ed      mov   edi, 0x40258c                        ; U"utf32le> \\u00a2\\u20ac\\U00010348 in cyan:.[36m ... .[0m."
+            0x004016ed      mov   edi, 0x40258c
             0x004016f7      mov   edi, 0x40266c                        ; U"Mountain range with embedded quad zeros: .A.A.A."

[XX] db/cmd/cmd_pd ANSI and wide string flag cmts
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
e str.escbslash=false
e asm.cmt.right=true
pd 1 @ 0x140001004
pd 1 @ 0x140001010
pd 1 @ 0x14000101c
e asm.cmt.right=false
pd 1 @ 0x140001004
pd 1 @ 0x140001010
pd 1 @ 0x14000101c
' bins/pe/testapp-msvc64.exe
-- stdout
--- expected
+++ actual
@@ -0,15 +0,13 @@
             0x140001004      488d05f54f01.  lea   rax, str.ANSI__esc:__e_33m ; section..data
                                                                        ; 0x140016000 ; "\tANSI\esc: \x1b[33m\r\n"
-            0x140001010      488d05015001.  lea   rax, str.wide__esc:__e_0m ; 0x140016018 ; u"\twide\esc: \x1b[0m\xa1\r\n"
-            0x14000101c      488d051d5001.  lea   rax, str.wide__in_Arabic: ; 0x140016040 ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
+            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; u"\twide\esc: \x1b[0m\xa1\r\n"
+            0x14000101c      488d051d5001.  lea   rax, [0x140016040]   ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
             ; section..data
             ; 0x140016000
             ; "\tANSI\esc: \x1b[33m\r\n"
             0x140001004      488d05f54f01.  lea rax, str.ANSI__esc:__e_33m
-            ; 0x140016018
             ; u"\twide\esc: \x1b[0m\xa1\r\n"
-            0x140001010      488d05015001.  lea rax, str.wide__esc:__e_0m
-            ; 0x140016040
+            0x140001010      488d05015001.  lea rax, [0x140016018]
             ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
-            0x14000101c      488d051d5001.  lea rax, str.wide__in_Arabic:
+            0x14000101c      488d051d5001.  lea rax, [0x140016040]

[XX] db/cmd/cmd_pd asm.noisy
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.sub.names=false
e asm.bytes=true
e asm.noisy=true
e str.escbslash=false
e asm.cmt.right=true
pd 1 @ 0x140001010
e asm.cmt.right=false
pd 1 @ 0x140001010
' bins/pe/testapp-msvc64.exe
-- stdout
--- expected
+++ actual
@@ -1,4 +1,3 @@
-            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; str.wide__esc:__e_0m ; u"\twide\esc: \x1b[0m\xa1\r\n"
-            ; str.wide__esc:__e_0m
+            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; u"\twide\esc: \x1b[0m\xa1\r\n"
             ; u"\twide\esc: \x1b[0m\xa1\r\n"
             0x140001010      488d05015001.  lea rax, [0x140016018]
```

It looks like it doesn't create a flag in the `str.` namespace.

borzacchiello commented:
> What worries me, are these regressions:
> 
> ```
> 
> [XX] db/cmd/cmd_pd bin.str.enc utf32le
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=false
> e asm.cmt.off=false
> e str.escbslash=true
> e bin.str.enc=utf16le
> pd 1 @ 0x004016de
> e bin.str.enc=utf32le
> pd 1 @ 0x004016de
> e asm.sub.names=false
> e asm.noisy=false
> e bin.str.enc=guess
> pd 1 @ 0x004016ed
> pd 1 @ 0x004016f7
> pd 1 @ 0x00401701
> e bin.str.enc=utf32le
> pd 1 @ 0x00401701
> ' bins/elf/strenc
> -- stdout
> --- expected
> +++ actual
> @@ -1,6 +1,6 @@
> -            0x004016de      mov   esi, str.Linux_wide__esc:__e_0m      ; u"\t"
> -            0x004016de      mov   esi, str.Linux_wide__esc:__e_0m      ; U"\tLinux_wide\\esc: \x1b[0m\xa1\r\n"
> -            0x004016ed      mov   edi, 0x40258c                        ; U"utf32le> \\u00a2\\u20ac\\U00010348 in cyan:\x1b[36m \xa2\u20ac\U00010348 \x1b[0m\n"
> +            0x004016de      mov   esi, 0x402528                        ; '(%@' ; u"\t"
> +            0x004016de      mov   esi, 0x402528                        ; '(%@' ; U"\tLinux_wide\\esc: \x1b[0m\xa1\r\n"
> +            0x004016ed      mov   edi, 0x40258c
> @@ -3,5 +3,5 @@
>              0x00401693      mov   edi, 0x4021ff                        ; "utf8> \\u00a2\\u20ac\\U00010348 in yellow:.[33m ......... .[0m\n"
>              0x00401693      mov   edi, 0x4021ff                        ; "utf8> \\u00a2\\u20ac\\U00010348 in yellow:.[33m ... .[0m."
>              0x004016ac      mov   edi, 0x40224a                        ; 'J"@' ; u"utf16le> \\u00a2\\u20ac\\U00010348 in green:.[32m ... .[0m."
> -            0x004016ed      mov   edi, 0x40258c                        ; U"utf32le> \\u00a2\\u20ac\\U00010348 in cyan:.[36m ... .[0m."
> +            0x004016ed      mov   edi, 0x40258c
>              0x004016f7      mov   edi, 0x40266c                        ; U"Mountain range with embedded quad zeros: .A.A.A."
> 
> [XX] db/cmd/cmd_pd ANSI and wide string flag cmts
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
> e str.escbslash=false
> e asm.cmt.right=true
> pd 1 @ 0x140001004
> pd 1 @ 0x140001010
> pd 1 @ 0x14000101c
> e asm.cmt.right=false
> pd 1 @ 0x140001004
> pd 1 @ 0x140001010
> pd 1 @ 0x14000101c
> ' bins/pe/testapp-msvc64.exe
> -- stdout
> --- expected
> +++ actual
> @@ -0,15 +0,13 @@
>              0x140001004      488d05f54f01.  lea   rax, str.ANSI__esc:__e_33m ; section..data
>                                                                         ; 0x140016000 ; "\tANSI\esc: \x1b[33m\r\n"
> -            0x140001010      488d05015001.  lea   rax, str.wide__esc:__e_0m ; 0x140016018 ; u"\twide\esc: \x1b[0m\xa1\r\n"
> -            0x14000101c      488d051d5001.  lea   rax, str.wide__in_Arabic: ; 0x140016040 ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
> +            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; u"\twide\esc: \x1b[0m\xa1\r\n"
> +            0x14000101c      488d051d5001.  lea   rax, [0x140016040]   ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
>              ; section..data
>              ; 0x140016000
>              ; "\tANSI\esc: \x1b[33m\r\n"
>              0x140001004      488d05f54f01.  lea rax, str.ANSI__esc:__e_33m
> -            ; 0x140016018
>              ; u"\twide\esc: \x1b[0m\xa1\r\n"
> -            0x140001010      488d05015001.  lea rax, str.wide__esc:__e_0m
> -            ; 0x140016040
> +            0x140001010      488d05015001.  lea rax, [0x140016018]
>              ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
> -            0x14000101c      488d051d5001.  lea rax, str.wide__in_Arabic:
> +            0x14000101c      488d051d5001.  lea rax, [0x140016040]
> 
> [XX] db/cmd/cmd_pd asm.noisy
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.sub.names=false
> e asm.bytes=true
> e asm.noisy=true
> e str.escbslash=false
> e asm.cmt.right=true
> pd 1 @ 0x140001010
> e asm.cmt.right=false
> pd 1 @ 0x140001010
> ' bins/pe/testapp-msvc64.exe
> -- stdout
> --- expected
> +++ actual
> @@ -1,4 +1,3 @@
> -            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; str.wide__esc:__e_0m ; u"\twide\esc: \x1b[0m\xa1\r\n"
> -            ; str.wide__esc:__e_0m
> +            0x140001010      488d05015001.  lea   rax, [0x140016018]   ; u"\twide\esc: \x1b[0m\xa1\r\n"
>              ; u"\twide\esc: \x1b[0m\xa1\r\n"
>              0x140001010      488d05015001.  lea rax, [0x140016018]
> ```

I'm looking at them. Is always the same issue, the strings are detected as big endian with the address shifted by one. For example, I'm currently looking at this one:
```
> +            0x14000101c      488d051d5001.  lea   rax, [0x140016040]   ; u"\"wide\" in Arabic: \u0648\u0627\u0633\u0639"
```
The problem is that we are adding the very same string at address `0x14001603f` as UTF16BE

borzacchiello committed:
Changed UTF BE heuristic

If there is any ambiguity about big-endian or little-endian, always
choose little endian

borzacchiello commented:
This should be almost ok. Now if there is any ambiguity between _little-endian_ and _big-endian_, we will always choose _little-endian_. This is not the ideal solution I think, but at least does not break old tests.

The remaining failing tests should regard only string counts. I will look into them.

XVilka commented on test/unit/test_str_search.c:
```diff
@@ -0,0 +1,189 @@
+	mu_run_test(test_rz_scan_strings_detect_ascii);
+	mu_run_test(test_rz_scan_strings_detect_utf8);
+	mu_run_test(test_rz_scan_strings_detect_utf16_le);
+	// mu_run_test(test_rz_scan_strings_detect_utf16_be);
```
Why these are commented out?

borzacchiello replied:
Because the buffers are ambiguous. For example, the UTF16 string can be either
`I am a UTF-16be string` UTF16-be starting at offset 3
or
`I am a UTF-16be strin` UTF16-le starting at offset 4

In the current state of the PR, the string is detected as the second one. I would like to change the test so that the buffer is non-ambiguous.

XVilka replied:
With my suggestion you could set the preferred option in these test to avoid ambiguity. I also recommend to add a couple more tests with the different preferred string endianness in the options.

XVilka commented:
There is also this, but it looks like false positive both before and after anyway:
```

[XX] db/cmd/cmd_i bin.str.enc string reset
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'mkdir .tmp
cp bins/elf/analysis/x86-helloworld-gcc .tmp/strenc-bin
o+ .tmp/strenc-bin
iz:quiet
s 0x080484b0
wz k\0i\0abcdefgh
(izenc enc; e bin.str.enc=$0; iz:quiet)
.(izenc guess)
.(izenc utf8)
.(izenc utf16le)
o-*
rm .tmp/strenc-bin
' --
-- stdout
--- expected
+++ actual
@@ -1,4 +1,4 @@
 0   0x000004b0 0x080484b0 12  13   .rodata ascii Hello world!
 0   0x000004b4 0x080484b4 8   9    .rodata ascii abcdefgh
 0   0x000004b4 0x080484b4 8   9    .rodata ascii abcdefgh
-0   0x000004b0 0x080484b0 6   13   .rodata utf16le ki扡摣晥桧 blocks=Basic Latin,CJK Unified Ideographs
+0   0x000004ab 0x080484ab 9   18   .rodata utf16le ĀȀ欀椀愀换敤杦h blocks=Latin Extended-A,Latin Extended-B,CJK Unified Ideographs,Basic Latin
```
Thus, I think it's not critical.

XVilka commented:
Regarding the ambiguity... I thought about adding a "preferred" endianess int the `RzUtilStrScanOptions` structure. When we scan real binary, where we know the default endianess, e.g. if the architecture is x86, more likely the default is LE, if the architecture is armv7 BE, the strings are likely BE as well. It will solve the ambiguity problem more gracefully, what do you think?
cc @ret2libc @wargio @thestr4ng3r

borzacchiello commented:
> Regarding the ambiguity... I thought about adding a "preferred" endianess int the `RzUtilStrScanOptions` structure. When we scan real binary, where we know the default endianess, e.g. if the architecture is x86, more likely the default is LE, if the architecture is armv7 BE, the strings are likely BE as well. It will solve the ambiguity problem more gracefully, what do you think?
> cc @ret2libc @wargio @thestr4ng3r

I think it is a good solution. I can implement it!

borzacchiello commented:
These are the existing debugged regressions:

---

## Regression 1
```
[XX] db/cmd/cmd_i ascii substring detection (#14499)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc izz~pomf bins/pe/Reborn_Stub-strings.exe
-- stdout
--- expected
+++ actual
@@ -1,4 +1,4 @@
-7154 0x00087f8a 0x00489d8a 26  53   .text   utf16le http://pomf.cat/upload.php
-7159 0x000880dd 0x00489edd 19  39   .text   utf16le https://a.pomf.cat/
-7322 0x00089bed 0x0048b9ed 26  53   .text   utf16le http://pomf.cat/upload.php
-7323 0x00089c22 0x0048ba22 19  40   .text   utf16le https://a.pomf.cat/
+7180 0x00087f8a 0x00489d8a 26  53   .text   utf16le http://pomf.cat/upload.php
+7185 0x000880dd 0x00489edd 19  39   .text   utf16le https://a.pomf.cat/
+7349 0x00089bed 0x0048b9ed 26  53   .text   utf16le http://pomf.cat/upload.php
+7350 0x00089c22 0x0048ba22 19  40   .text   utf16le https://a.pomf.cat/
```

We find more strings. Diff between the outputs of `izz` on `bins/pe/Reborn_Stub-strings.exe` (rizin before this PR vs rizin after this PR):
```diff
3183a3183
> 0x0002760c 0x0042940c 6   11   .text   utf8    Ny䊈ϗN۬ blocks=Basic Latin,CJK Unified Ideographs Extension A,Greek and Coptic,Arabic
4604a4605
> 0x00039331 0x0043b131 4   5    .text   ascii   \a\bo4
4748a4750
> 0x0003adc7 0x0043cbc7 4   20   .text   utf32be Æ}/8
4805a4808
> 0x0003b44a 0x0043d24a 4   20   .text   utf32be 
4932a4936
> 0x0003c700 0x0043e500 4   20   .text   utf32be ].7Ç
4976c4980,4981
< 0x0003cc4f 0x0043ea4f 5   24   .text   utf32le 7(YË blocks=Basic Latin,Latin-1 Supplement,Private Use Area
---
> 0x0003cc3c 0x0043ea3c 4   20   .text   utf32be ¡u¹¿
> 0x0003cc53 0x0043ea53 4   20   .text   utf32le (YË blocks=Basic Latin,Latin-1 Supplement,Private Use Area
5030a5036
> 0x0003d488 0x0043f288 4   20   .text   utf32be <­{Ð
5050a5057
> 0x0003db79 0x0043f979 4   20   .text   utf32be 8Ïm8
5189c5196
< 0x0003f6fb 0x004414fb 5   24   .text   utf32le \e<ŊĔ\ blocks=Basic Latin,Latin Extended-A
---
> 0x0003f6f4 0x004414f4 4   20   .text   utf32be Ð\e<J
5242c5249
< 0x000401af 0x00441faf 4   5    .text   ascii   +U\ar
---
> 0x0004019c 0x00441f9c 4   20   .text   utf32be #GWÔ
5304a5312
> 0x00040e31 0x00442c31 4   20   .text   utf32be QjúÇ
5579a5588
> 0x000448c8 0x004466c8 4   20   .text   utf32be T»ÖÆ
5586c5595
< 0x000449ef 0x004467ef 4   20   .text   utf32le !7p딸 blocks=Basic Latin,Hangul Syllables
---
> 0x000449e8 0x004467e8 5   24   .text   utf32be º!7p8
5633c5642
< 0x0004513c 0x00446f3c 7   9    .text   utf8    \b \aZْ%+ blocks=Basic Latin,Arabic
---
> 0x0004513b 0x00446f3b 8   10   .text   utf8    ,\b \aZْ%+ blocks=Basic Latin,Arabic
5750a5760
> 0x00046844 0x00448644 4   20   .text   utf32be 2FQÓ
5756a5767
> 0x00046914 0x00448714 5   24   .text   utf32be  ÃQwÆ
5795d5805
< 0x000472ea 0x004490ea 10  44   .text   utf32le n;şÁłƩųWØČ blocks=Basic Latin,Latin Extended-A,Latin-1 Supplement,Latin Extended-B
5891d5900
< 0x000488cc 0x0044a6cc 7   32   .text   utf32le kõUŤŊƚ¼ blocks=Basic Latin,Latin-1 Supplement,Latin Extended-A,Latin Extended-B
5999c6008
< 0x00049f8c 0x0044bd8c 4   20   .text   utf32le 1Qg¤ blocks=Basic Latin,Latin-1 Supplement
---
> 0x00049f85 0x0044bd85 5   24   .text   utf32be Ä1Qg¤
6043a6053
> 0x0004a84b 0x0044c64b 4   20   .text   utf32be Zöt8
6103a6114
> 0x0004b62a 0x0044d42a 4   20   .text   utf32be "HjÐ
6221a6233,6236
> 0x0004dcd9 0x0044fad9 6   14   .text   utf16be -#:\rĐ÷
> 0x0004dcfd 0x0044fafd 7   16   .text   utf16be -(D\rĐ(㬀
> 0x0004de9b 0x0044fc9b 4   10   .text   utf16be É^ó\n
> 0x0004dead 0x0044fcad 4   10   .text   utf16be -aó\n
6223a6239
> 0x00055615 0x00457415 12  26   .text   utf16be $Å%Å&ÅCÅHÅ\tR
6225a6242
> 0x0005760d 0x0045940d 5   12   .text   utf16be W\a\b\eȈ
6305a6323,6327
> 0x0005f05d 0x00460e5d 16  17   .text   ascii   DelegateCallback
> 0x0005f06e 0x00460e6e 18  19   .text   ascii   DelegateAsyncState
> 0x0005f081 0x00460e81 9   10   .text   ascii   EndInvoke
> 0x0005f08b 0x00460e8b 19  20   .text   ascii   DelegateAsyncResult
> 0x0005f09f 0x00460e9f 6   7    .text   ascii   Invoke
7086a7109,7111
> 0x00087402 0x00489202 8   17   .text   utf16le egui.exe
> 0x00087415 0x00489215 9   20   .text   utf16be vscan.exe
> 0x0008742b 0x0048922b 73  147  .text   utf16le Software\Microsoft\Windows NT\CurrentVersion\Image File Execution Options
7302c7327,7328
< 0x00089144 0x0048af44 100 201  .text   utf16le }{0}Host: {2}{0}Username: {3}{0}Password: {4}{0}================================================={0}
---
> 0x000890a4 0x0048aea4 11  24   .text   utf16be stext "{0}"
> 0x000890be 0x0048aebe 167 335  .text   utf16le ================================================={0}Application: {1}{0}Host: {2}{0}Username: {3}{0}Password: {4}{0}================================================={0}
7444,7445c7470,7471
< 0x0008ba46 0x0048e646 13  28   .rsrc   utf16le _VERSION_INFO
< 0x0008baa0 0x0048e6a0 10  22   .rsrc   utf16le arFileInfo
---
> 0x0008ba42 0x0048e642 15  32   .rsrc   utf16le VS_VERSION_INFO
> 0x0008ba9e 0x0048e69e 11  24   .rsrc   utf16le VarFileInfo
```

Most of the new strings are false positives, but some of them are not (e.g. `DelegateCallback` at address `0x00460e5d`). This is due to an issue in the code before this PR:
if you look at [this](https://github.com/rizinorg/rizin/blob/9377b16951b8f164081de0ecb41b59059c52c2ae/librz/bin/bfile.c#L288) line of code, the detection of the current string failed because the string considered a false positive. The problem is that needle is not reset, so the next iteration would skip some bytes.
This is the hex dump near `DelegateCallback` string:
```
[0x0048d12e]> s 0x00460e40
[0x00460e40]> px
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00460e40  0041 7379 6e63 4361 6c6c 6261 636b 0061  .AsyncCallback.a
0x00460e50  6464 7265 7373 0078 0079 007a 0044 656c  ddress.x.y.z.Del
0x00460e60  6567 6174 6543 616c 6c62 6163 6b00 4465  egateCallback.De
```
the old code would start the detection of a utf16-le string at address `0x00460e56`, build the string `xyz敄敬慧整慃汬慢正䐀汥 [...]` (it eats the bytes after `xyz` interpreting them as utf16-le), and then discard it during the false positive check. But since `needle` is not reset, `DelegateCallback` is lost.

---

## Regression 2

```
[XX] db/cmd/cmd_i regression for #9370
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -m 0x10000 -Qc izz~Hello bins/elf/analysis/hello-arm32
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-6   0x00000200 0x00010200 11  12   .rodata         ascii   Hello World
+7   0x00000200 0x00010200 11  12   .rodata         ascii   Hello World
```

We find one additional string. This is the output of diff wrt the command `izz` on `bins/elf/analysis/hello-arm32`

```diff
>   0x00000055 0x00000055 4   20                   utf32be ÔÔ𐃔𐀑
```

while it is a false positive, the string exists at this offset.

---

## Regression 3

```
[XX] db/cmd/cmd_i bin.str.enc string reset
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'mkdir .tmp
cp bins/elf/analysis/x86-helloworld-gcc .tmp/strenc-bin
o+ .tmp/strenc-bin
iz:quiet
s 0x080484b0
wz k\0i\0abcdefgh
(izenc enc; e bin.str.enc=$0; iz:quiet)
.(izenc guess)
.(izenc utf8)
.(izenc utf16le)
o-*
rm .tmp/strenc-bin
' --
-- stdout
--- expected
+++ actual
@@ -1,4 +1,4 @@
 0   0x000004b0 0x080484b0 12  13   .rodata ascii Hello world!
 0   0x000004b4 0x080484b4 8   9    .rodata ascii abcdefgh
 0   0x000004b4 0x080484b4 8   9    .rodata ascii abcdefgh
-0   0x000004b0 0x080484b0 6   13   .rodata utf16le ki扡摣晥桧 blocks=Basic Latin,CJK Unified Ideographs
+0   0x000004ab 0x080484ab 9   18   .rodata utf16le ĀȀ欀椀愀换敤杦h blocks=Latin Extended-A,Latin Extended-B,CJK Unified Ideographs,Basic Latin
```

These two strings seems to be false positives. With the PR the string is detected at an earlier address due to a similar issue wrt the `DelegateCallback` case described in regression 1.

---

## Regression 4

```
[XX] db/formats/mdmp mdmp 64bit - strings count
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iz~? bins/mdmp/hello64.dmp
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-22900
+22915
```

This is the diff of the `iz` command on `bins/mdmp/hello64.dmp`:
```diff
156a157,159
>  0x002e2fe1 0x77820352 4 10 C:_Windows_System32__tdll.dll utf16le Alpc
>  0x002e2feb 0x7782035c 4 10 C:_Windows_System32__tdll.dll utf16le Pool
>  0x002e2ff5 0x77820366 5 12 C:_Windows_System32__tdll.dll utf16le Timer
221a225
>  0x002e7e0e 0x7782517f 96 194 C:_Windows_System32__tdll.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
2214c2218,2220
<  0x002f38f3 0x77830c64 22 92 C:_Windows_System32__tdll.dll utf32le @@3;;;;њȽȾȿɀɂ|VɃmϱϸɄϭў blocks=Basic Latin,Cyrillic,Latin Extended-B,Greek and Coptic
---
>  0x002f38c0 0x77830c31 4 20 C:_Windows_System32__tdll.dll utf32be ÁÁÁ8
>  0x002f38f4 0x77830c65 7 32 C:_Windows_System32__tdll.dll utf32be @3;;;;Z
>  0x002f3927 0x77830c98 9 40 C:_Windows_System32__tdll.dll utf32le |VɃmϱϸɄϭў blocks=Basic Latin,Latin Extended-B,Greek and Coptic,Cyrillic
2218a2225
>  0x002f3e8a 0x778311fb 6 28 C:_Windows_System32__tdll.dll utf32be ÝÞßàáâ
2227a2235
>  0x002f4ca6 0x77832017 4 10 C:_Windows_System32__tdll.dll utf16be ê777
2974c2982
<  0x00330d89 0x7786e0fa 13 28 C:_Windows_System32__tdll.dll utf16le _VERSION_INFO
---
>  0x00330d85 0x7786e0f6 15 32 C:_Windows_System32__tdll.dll utf16le VS_VERSION_INFO
2995c3003
<  0x003310c3 0x7786e434 10 22 C:_Windows_System32__tdll.dll utf16le arFileInfo
---
>  0x003310c1 0x7786e432 11 24 C:_Windows_System32__tdll.dll utf16le VarFileInfo
5209c5217
<  0x00161a3b 0x7759ddac 5 12 C:_Windows_System32_kernel32.dll utf16le n.ini
---
>  0x00161a37 0x7759dda8 7 16 C:_Windows_System32_kernel32.dll utf16le win.ini
5219c5227
<  0x00161c83 0x7759dff4 18 38 C:_Windows_System32_kernel32.dll utf16le lasses\Wow6432Node
---
>  0x00161c7f 0x7759dff0 20 42 C:_Windows_System32_kernel32.dll utf16le \Classes\Wow6432Node
5221c5229
<  0x00161dc3 0x7759e134 43 88 C:_Windows_System32_kernel32.dll utf16le EGISTRY\USER\*\SOFTWARE\Classes\Wow6432Node
---
>  0x00161dbf 0x7759e130 45 92 C:_Windows_System32_kernel32.dll utf16le \REGISTRY\USER\*\SOFTWARE\Classes\Wow6432Node
5324a5333
>  0x0016697a 0x775a2ceb 96 194 C:_Windows_System32_kernel32.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
15136c15145
<  0x001a7967 0x775e3cd8 4 5 C:_Windows_System32_kernel32.dll ascii 0MTw
---
> 0x001a7967 0x775e3cd8 4 5 C:_Windows_System32_kernel32.dll ascii 0MTw
17013a17023
> 0x001b729c 0x775f360d 4 20 C:_Windows_System32_kernel32.dll utf32be ±Ïì\n
17096c17106
< 0x001b8063 0x775f43d4 12 26 C:_Windows_System32_kernel32.dll utf16le \[]:|<>+=;,?
---
> 0x001b805f 0x775f43d0 14 30 C:_Windows_System32_kernel32.dll utf16le "/\[]:|<>+=;,?
17185a17196
> 0x001b8e93 0x775f5204 6 7 C:_Windows_System32_kernel32.dll ascii BAUD=#
17202c17213
< 0x001b8f4b 0x775f52bc 5 12 C:_Windows_System32_kernel32.dll utf16le inSxs
---
> 0x001b8f47 0x775f52b8 7 16 C:_Windows_System32_kernel32.dll utf16le \WinSxs
18737c18748
<  0x001d9d49 0x776160ba 13 28 C:_Windows_System32_kernel32.dll utf16le _VERSION_INFO
---
>  0x001d9d45 0x776160b6 15 32 C:_Windows_System32_kernel32.dll utf16le VS_VERSION_INFO
18758c18769
<  0x001da0a7 0x77616418 10 22 C:_Windows_System32_kernel32.dll utf16le arFileInfo
---
>  0x001da0a5 0x77616416 11 24 C:_Windows_System32_kernel32.dll utf16le VarFileInfo
18807a18819
>  0x003dee8e 0x7fefd4ec1ff 96 194 C:_Windows_System32_KERNELBASE.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
20290c20302
<  0x003fad49 0x7fefd5080ba 13 28 C:_Windows_System32_KERNELBASE.dll utf16le _VERSION_INFO
---
>  0x003fad45 0x7fefd5080b6 15 32 C:_Windows_System32_KERNELBASE.dll utf16le VS_VERSION_INFO
20311c20323
<  0x003fb0af 0x7fefd508420 10 22 C:_Windows_System32_KERNELBASE.dll utf16le arFileInfo
---
>  0x003fb0ad 0x7fefd50841e 11 24 C:_Windows_System32_KERNELBASE.dll utf16le VarFileInfo
20373a20386
>  0x00479c90 0x7fefdb2d001 19 40 C:_Windows_System32_msvcrt.dll utf16be \b¬\b¹\bÈ\bÐ\bÚ\bå\bò\bý\b\nĈ
20375a20389
>  0x0047aa5a 0x7fefdb2ddcb 96 194 C:_Windows_System32_msvcrt.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
21867a21882
>  0x00480857 0x7fefdb33bc8 4 10 C:_Windows_System32_msvcrt.dll utf16le PATH
21874c21889
<  0x004808cb 0x7fefdb33c3c 8 18 C:_Windows_System32_msvcrt.dll utf16le stemRoot
---
>  0x004808c7 0x7fefdb33c38 10 22 C:_Windows_System32_msvcrt.dll utf16le SystemRoot
22850c22865
<  0x0049155b 0x7fefdb448cc 4 20 C:_Windows_System32_msvcrt.dll utf32le \v§\r· blocks=Basic Latin,Latin-1 Supplement
---
>  0x00491554 0x7fefdb448c5 5 24 C:_Windows_System32_msvcrt.dll utf32be ¤\v§\r·
22878c22893
<  0x00499cf9 0x7fefdb4d06a 13 28 C:_Windows_System32_msvcrt.dll utf16le _VERSION_INFO
---
>  0x00499cf5 0x7fefdb4d066 15 32 C:_Windows_System32_msvcrt.dll utf16le VS_VERSION_INFO
22899c22914
<  0x0049a047 0x7fefdb4d3b8 10 22 C:_Windows_System32_msvcrt.dll utf16le arFileInfo
---
>  0x0049a045 0x7fefdb4d3b6 11 24 C:_Windows_System32_msvcrt.dll utf16le VarFileInfo
```

Similar to regression 1

---

## Regression 5

```
[XX] db/formats/mdmp mdmp 32bit - strings count
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iz~? bins/mdmp/hello.dmp
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-10276
+10300
```

This is the diff of the `iz` command on `bins/mdmp/hello.dmp`:
```diff
155a156,158
>  0x0043cf24 0x77820352 4 10 C:_Windows_System32__tdll.dll utf16le Alpc
>  0x0043cf2e 0x7782035c 4 10 C:_Windows_System32__tdll.dll utf16le Pool
>  0x0043cf38 0x77820366 5 12 C:_Windows_System32__tdll.dll utf16le Timer
220a224
>  0x00441d51 0x7782517f 96 194 C:_Windows_System32__tdll.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
1044,1047c1048,1051
<  0x00447247 0x7782a675 35 36 C:_Windows_System32__tdll.dll ascii RtlFindActivationContextSectionGuid
<  0x0044726b 0x7782a699 37 38 C:_Windows_System32__tdll.dll ascii RtlFindActivationContextSectionString
<  0x00447291 0x7782a6bf 26 27 C:_Windows_System32__tdll.dll ascii RtlFindCharInUnicodeString
<  0x004472ac 0x7782a6da 16 17 C:_Windows_System32__tdll.dll ascii RtlFindClearBits
---
> 0x00447247 0x7782a675 35 36 C:_Windows_System32__tdll.dll ascii RtlFindActivationContextSectionGuid
> 0x0044726b 0x7782a699 37 38 C:_Windows_System32__tdll.dll ascii RtlFindActivationContextSectionString
> 0x00447291 0x7782a6bf 26 27 C:_Windows_System32__tdll.dll ascii RtlFindCharInUnicodeString
> 0x004472ac 0x7782a6da 16 17 C:_Windows_System32__tdll.dll ascii RtlFindClearBits
2213c2217,2219
< 0x0044d836 0x77830c64 22 92 C:_Windows_System32__tdll.dll utf32le @@3;;;;њȽȾȿɀɂ|VɃmϱϸɄϭў blocks=Basic Latin,Cyrillic,Latin Extended-B,Greek and Coptic
---
> 0x0044d803 0x77830c31 4 20 C:_Windows_System32__tdll.dll utf32be ÁÁÁ8
> 0x0044d837 0x77830c65 7 32 C:_Windows_System32__tdll.dll utf32be @3;;;;Z
> 0x0044d86a 0x77830c98 9 40 C:_Windows_System32__tdll.dll utf32le |VɃmϱϸɄϭў blocks=Basic Latin,Latin Extended-B,Greek and Coptic,Cyrillic
2217a2224
> 0x0044ddcd 0x778311fb 6 28 C:_Windows_System32__tdll.dll utf32be ÝÞßàáâ
2226a2234
> 0x0044ebe9 0x77832017 4 10 C:_Windows_System32__tdll.dll utf16be ê777
2971c2979
<  0x0048accc 0x7786e0fa 13 28 C:_Windows_System32__tdll.dll utf16le _VERSION_INFO
---
>  0x0048acc8 0x7786e0f6 15 32 C:_Windows_System32__tdll.dll utf16le VS_VERSION_INFO
2992c3000
<  0x0048b006 0x7786e434 10 22 C:_Windows_System32__tdll.dll utf16le arFileInfo
---
>  0x0048b004 0x7786e432 11 24 C:_Windows_System32__tdll.dll utf16le VarFileInfo
5148c5156
<  0x00168c3c 0x7500d06a 13 28 C:_Windows_System32_wow64.dll utf16le _VERSION_INFO
---
>  0x00168c38 0x7500d066 15 32 C:_Windows_System32_wow64.dll utf16le VS_VERSION_INFO
5169c5177
<  0x00168f82 0x7500d3b0 10 22 C:_Windows_System32_wow64.dll utf16le arFileInfo
---
>  0x00168f80 0x7500d3ae 11 24 C:_Windows_System32_wow64.dll utf16le VarFileInfo
5200c5208
<  0x00128c3c 0x74fc906a 13 28 C:_Windows_System32_wow64win.dll utf16le _VERSION_INFO
---
>  0x00128c38 0x74fc9066 15 32 C:_Windows_System32_wow64win.dll utf16le VS_VERSION_INFO
5221c5229
<  0x00128faa 0x74fc93d8 10 22 C:_Windows_System32_wow64win.dll utf16le arFileInfo
---
>  0x00128fa8 0x74fc93d6 11 24 C:_Windows_System32_wow64win.dll utf16le VarFileInfo
5223c5231
<  0x000cdc3c 0x74f6606a 13 28 C:_Windows_System32_wow64cpu.dll utf16le _VERSION_INFO
---
>  0x000cdc38 0x74f66066 15 32 C:_Windows_System32_wow64cpu.dll utf16le VS_VERSION_INFO
5244c5252
<  0x000cdf86 0x74f663b4 10 22 C:_Windows_System32_wow64cpu.dll utf16le arFileInfo
---
>  0x000cdf84 0x74f663b2 11 24 C:_Windows_System32_wow64cpu.dll utf16le VarFileInfo
7398a7407
> 0x00628d2f 0x77a4215d 95 192 C:_Windows_SysWOW64__tdll.dll utf16be ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
7475c7484,7486
< 0x00638f88 0x77a523b6 8 9 C:_Windows_SysWOW64__tdll.dll ascii c%Q%W%]%
---
> 0x00638f17 0x77a52345 30 62 C:_Windows_SysWOW64__tdll.dll utf16be üéâäàåçêëèïîìÄÅÉæÆôöòûùÿÖÜø£Ø×
> 0x00638f57 0x77a52385 15 32 C:_Windows_SysWOW64__tdll.dll utf16be íóúñÑªº¿®¬½¼¡«»
> 0x00638f81 0x77a523af 9 20 C:_Windows_SysWOW64__tdll.dll utf16be ÂÀ©c║╗╝▢¥
7477a7489,7491
> 0x00638fb5 0x77a523e3 9 20 C:_Windows_SysWOW64__tdll.dll utf16be ðÐÊËÈ1ǍÎÏ
> 0x00638fd7 0x77a52405 17 36 C:_Windows_SysWOW64__tdll.dll utf16be ßÔÒõÕµþÞÚÛÙýÝ¯´­±
> 0x00638ffd 0x77a5242b 12 26 C:_Windows_SysWOW64__tdll.dll utf16be ¶§÷¸°¨·¹³² ■
7481c7495,7497
< 0x0063918a 0x77a525b8 8 9 C:_Windows_SysWOW64__tdll.dll ascii c%Q%W%]%
---
> 0x00639119 0x77a52547 30 62 C:_Windows_SysWOW64__tdll.dll utf16be üéâäàåçêëèïîìÄÅÉæÆôöòûùÿÖÜø£Ø×
> 0x00639159 0x77a52587 15 32 C:_Windows_SysWOW64__tdll.dll utf16be íóúñÑªº¿®¬½¼¡«»
> 0x00639183 0x77a525b1 9 20 C:_Windows_SysWOW64__tdll.dll utf16be ÂÀ©c║╗╝▢¥
7483a7500,7502
> 0x006391b7 0x77a525e5 9 20 C:_Windows_SysWOW64__tdll.dll utf16be ðÐÊËÈ1ǍÎÏ
> 0x006391d9 0x77a52607 17 36 C:_Windows_SysWOW64__tdll.dll utf16be ßÔÒõÕµþÞÚÛÙýÝ¯´­±
> 0x006391ff 0x77a5262d 12 26 C:_Windows_SysWOW64__tdll.dll utf16be ¶§÷¸°¨·¹³² ■
7647a7667
>  0x0029591f 0x77083d4d 97 196 C:_Windows_SysWOW64_kernel32.dll utf16be  ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛ+ÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ
8596c8616
<  0x00299106 0x77087534 11 12 C:_Windows_SysWOW64_kernel32.dll ascii _vfprintf_l
---
> 0x00299106 0x77087534 11 12 C:_Windows_SysWOW64_kernel32.dll ascii _vfprintf_l
9553c9573
<  0x002a5180 0x770935ae 5 12 C:_Windows_SysWOW64_kernel32.dll utf16le null)
---
>  0x002a517e 0x770935ac 6 14 C:_Windows_SysWOW64_kernel32.dll utf16le (null)
10073a10094
>  0x002b104e 0x7709f47c 7 11 C:_Windows_SysWOW64_kernel32.dll utf8 E؉EƋE܉E blocks=Basic Latin,Arabic,Latin Extended-B,Syriac
10075a10097
>  0x002b1282 0x7709f6b0 7 11 C:_Windows_SysWOW64_kernel32.dll utf8 E؉EƋE܉E blocks=Basic Latin,Arabic,Latin Extended-B,Syriac
10104c10126
<  0x001adc8c 0x75d930ba 13 28 C:_Windows_SysWOW64_KERNELBASE.dll utf16le _VERSION_INFO
---
>  0x001adc88 0x75d930b6 15 32 C:_Windows_SysWOW64_KERNELBASE.dll utf16le VS_VERSION_INFO
10125c10147
<  0x001adff2 0x75d93420 10 22 C:_Windows_SysWOW64_KERNELBASE.dll utf16le arFileInfo
---
>  0x001adff0 0x75d9341e 11 24 C:_Windows_SysWOW64_KERNELBASE.dll utf16le VarFileInfo
10181a10204,10205
>  0x00333b58 0x77412f86 14 19 C:_Windows_SysWOW64_msvcrt.dll utf8 7wԣ7wȣ7wģ7w#7w blocks=Basic Latin,Cyrillic Supplement,Latin Extended-B,Latin Extended-A
>  0x00333bf0 0x7741301e 11 15 C:_Windows_SysWOW64_msvcrt.dll utf8 7wԢ7w̢7wĢ7w blocks=Basic Latin,Cyrillic Supplement,Combining Diacritical Marks,Latin Extended-A
10209c10233
<  0x00335efe 0x7741532c 4 20 C:_Windows_SysWOW64_msvcrt.dll utf32le \v§\r· blocks=Basic Latin,Latin-1 Supplement
---
>  0x00335ef7 0x77415325 5 24 C:_Windows_SysWOW64_msvcrt.dll utf32be ¤\v§\r·
10217c10241
<  0x0033641e 0x7741584c 35 36 C:_Windows_SysWOW64_msvcrt.dll ascii rtion failed: %s, file %s, line %d\n
---
>  0x0033641a 0x77415848 39 40 C:_Windows_SysWOW64_msvcrt.dll ascii Assertion failed: %s, file %s, line %d\n
10254c10278
<  0x00337c3c 0x7741706a 13 28 C:_Windows_SysWOW64_msvcrt.dll utf16le _VERSION_INFO
---
>  0x00337c38 0x77417066 15 32 C:_Windows_SysWOW64_msvcrt.dll utf16le VS_VERSION_INFO
10275c10299
<  0x00337f8a 0x774173b8 10 22 C:_Windows_SysWOW64_msvcrt.dll utf16le arFileInfo
---
>  0x00337f88 0x774173b6 11 24 C:_Windows_SysWOW64_msvcrt.dll utf16le VarFileInfo
```

Similar to regression 1

---

As a final note, I think that regressions 1 and 2 should not check the ordinal number of the string: the regression is independent from the number of strings found in the binary. Is there any way for excluding the first column from the output of `izz`? It should simulate something like `izz | cut -d' ' -f1 --complement`.

XVilka commented:
@borzacchiello ok, the new output in all cases is better. What I suggest is to fix all these tests to not count the number of strings, but check the presense of these valid strings like:
```
0x001adff0 0x75d9341e 11 24 C:_Windows_SysWOW64_KERNELBASE.dll utf16le VarFileInfo
0x0044726b 0x7782a699 37 38 C:_Windows_System32__tdll.dll ascii RtlFindActivationContextSectionString
0x001b8f47 0x775f52b8 7 16 C:_Windows_System32_kernel32.dll utf16le \WinSxs
0x00161dbf 0x7759e130 45 92 C:_Windows_System32_kernel32.dll utf16le \REGISTRY\USER\*\SOFTWARE\Classes\Wow6432Node
0x0005f05d 0x00460e5d 16  17   .text   ascii   DelegateCallback
0x0005f06e 0x00460e6e 18  19   .text   ascii   DelegateAsyncState
0x0005f081 0x00460e81 9   10   .text   ascii   EndInvoke
0x0005f08b 0x00460e8b 19  20   .text   ascii   DelegateAsyncResult
0x0005f09f 0x00460e9f 6   7    .text   ascii   Invoke
...
```
Ignore the false positives in these tests and just add most of the true positives in the check instead.

After fixing the tests this PR should be mergeable then. Good job!

XVilka commented on test/unit/test_str_search.c:
```diff
@@ -0,0 +1,189 @@
+
```
Unnecessary empty line

wargio commented:
> Regarding the ambiguity... I thought about adding a "preferred" endianess int the `RzUtilStrScanOptions` structure. When we scan real binary, where we know the default endianess, e.g. if the architecture is x86, more likely the default is LE, if the architecture is armv7 BE, the strings are likely BE as well. It will solve the ambiguity problem more gracefully, what do you think?
> cc @ret2libc @wargio @thestr4ng3r

I think is definitely a good idea. already your code had drastically improved the string detection.

borzacchiello committed:
Small format fixes in str_search

borzacchiello committed:
Added preferred endianess switch in RzUtilStrScanOptions

borzacchiello committed:
Fixed str_search regression tests

XVilka commented:
Thank you, @borzacchiello!
